src/rdkafka_metadata.c

```diff
@@ -1971,3 +1971,179 @@ rd_kafka_metadata_new_topic_with_partition_replicas_mock(int replication_factor,
+                                            rk, METADATA, "METADATAUPDATE",
+                                            "Partition %s [%d]: leader epoch "
+                                            "is "
+                                            "not newer %" PRId32 " >= %" PRId32,
+                                            rd_kafka_Uuid_base64str(&topic_id),
+                                            part, current_leader_epoch,
+                                            mdpi->leader_epoch);
+                                } else {
+                                        rd_kafka_dbg(
+                                            rk, METADATA, "METADATAUPDATE",
+                                            "Partition %s [%d]: leader epoch "
+                                            "is "
+                                            "not newer %" PRId32 " >= %" PRId32,
+                                            topic, part, current_leader_epoch,
+                                            mdpi->leader_epoch);
+                                }
+                                continue;
+                        }
+                        cache_updated = rd_true;
+                        partition_cache_changes++;
+
+                        /* Need to acquire the write lock to avoid dirty reads
+                         * from other threads acquiring read locks. */
+                        rd_kafka_wrlock(rk);
+                        rkmce->rkmce_metadata_internal_topic.partitions[mdp->id]
+                            .leader_epoch = mdpi->leader_epoch;
+                        rkmce->rkmce_mtopic.partitions[mdp->id].leader =
+                            mdp->leader;
+                        rd_kafka_wrunlock(rk);
+                        rd_kafka_broker_destroy(rkb);
+
+                        if (by_id) {
+                                rd_kafka_dbg(rk, METADATA, "METADATAUPDATE",
+                                             "Partition %s [%d] "
+                                             " updated with leader %" PRId32
+                                             " and epoch %" PRId32,
+                                             rd_kafka_Uuid_base64str(&topic_id),
+                                             part, mdp->leader,
+                                             mdpi->leader_epoch);
+                        } else {
+                                rd_kafka_dbg(rk, METADATA, "METADATAUPDATE",
+                                             "Partition %s [%d] "
+                                             " updated with leader %" PRId32
+                                             " and epoch %" PRId32,
+                                             topic, part, mdp->leader,
+                                             mdpi->leader_epoch);
+                        }
+                }
+
+                if (partition_cache_changes > 0)
+                        rd_kafka_topic_metadata_update2(
```

Comment:
You can set `cache_updated` here instead of at each iteration
```suggestion
                if (partition_cache_changes > 0) {
                        cache_updated = rd_true;
                        rd_kafka_topic_metadata_update2(
```

Reply:
`cache_updated` is already being set as true above when `partition_cache_changes` is incremented

Reply:
Yes, but for each partition instead it can be set for each topic here

Reply:
Updated